Provide separate options to set the various optional tags

gnprice opened the pull request:
I find the `mm_tooLong` and `mm_badLength` tags super noisy, because
they appear on tons of notes, so I appreciate the option to disable
them.  But I'd like to do so while still setting the `mm_priority`
tag, which by definition only appears on notes useful for a priority
morpheme, which isn't so common.

Here, provide separate toggles for the three optional tags.  For a
user with the omnibus `Option_SetNotRequiredTags` setting in their
config -- including any user running this version for the first time
after using the previous version of MorphMan -- we migrate that
setting to the three new settings, giving exactly the same behavior.

For new users without an existing config, set the default to silence
the two noisy tags while still setting `mm_priority`.

One harmless but confusing gotcha when testing this change: if you
disable a tag and recalculate, the tag may still exist on some notes
if you have the `only update k+2 and below` setting set (as I do.)
This is probably the right behavior, in keeping with the general point
of that setting.

The PreferencesDialog code gets a very small refactor to make
`self.boolOptionList` shared across all the frames/tabs of the
dialog.  I like how very generic and crisp the relevant logic in
`readConfigFromGui` is, and how this change doesn't have to touch
that method at all to introduce the new checkboxes.

kaegi commented:
I'm reluctant to merge this change... The GUI is supposed to have as few options as possible, so that new users don't have to guess what is important and what not. I don't see a real benefit from disabling the tags. Or in other words: the disabling of the tags just for improving "database cosmetics" worsens the "GUI cosmetics" - which wins in this case. By that reasoning I also think adding the "Set tags even if not required option" is not really necessary, and wouldn't really oppose removing it.

If there's a pragmatic reason for having fewer tags, let me know.

gnprice commented:
Hmm, yeah, I definitely sympathize with wanting to reduce GUI options.

The original reason I made this change -- originally a very crude one-line version, which I cleaned up and added the configuration for this weekend in the course of starting to send changes as PRs -- was actually from my experience as a user first beginning to use MorphMan, a couple of months into using Anki, when a friend helped me import a large set of example sentences, get MorphMan set up, and start experimenting with using it on those cards. The `mm_badLength` and especially `mm_tooLong` tags were very confusing as I browsed through the cards -- what do these mean? Why do some cards have one, some the other, and some both? Why are some `mm_tooLong` cards apparently not penalized for length, while others are? The answers didn't make a lot of sense.

So within the first hour of trying MorphMan, I think, I'd knocked out the `mm_tooLong` tag -- just taken the assignment in `updateNotes` where it sets its local `tooLong`, and hard-coded that to False. More usefully, I'd adjusted the sentence-length settings, basically the rest of the changes that became PR #16; that line was right next to those changes, so it was easy to do.

Looking back at it, the most egregious aspects were things I fixed in #16 -- now `mm_tooLong` is only applied if the text is actually too long, in the sense that we're going to penalize it for being too long. So maybe it's fine now. I'll switch back to running with these tags added, and see if it bothers me again.

gnprice commented:
One minor thing does occur to me: it's kind of silly that most of the notes these tags apply to get two tags applied, `mm_badLength` and `mm_tooLong`. I think it'd be an improvement to make, say, an `mm_tooShort` tag and drop `mm_badLength`. Then it's still easy to filter for notes that are either too long or too short (just use `OR` in the query), and it's simpler to search for simply too short (no `-` necessary), but we simplify what the user sees on a given note by not showing two tags with similar meanings.

Also the name `mm_badLength` is kind of confusing in itself -- it isn't specific about what it means, and at least to me as a programmer that phrase "bad length" sounds like an error message, like something's corrupt or broken, not like some perfectly ordinary text just being a little long and given a lower priority as a result. So I'd be glad to replace it with something more specific.

kaegi commented:
> Also the name mm_badLength is kind of confusing in itself -- it isn't specific about what it means, and at least to me as a programmer that phrase "bad length" sounds like an error message, like something's corrupt or broken, not like some perfectly ordinary text just being a little long and given a lower priority as a result. So I'd be glad to replace it with something more specific.

Yes, that's exactly what I'm thinking too and be glad to accept a PR.

gnprice commented:
You got it! #23.
